File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/ui/screens/members/MembersViewModel.kt
package com.arcinteractive.spaces.ui.screens.members

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arcinteractive.spaces.data.auth.AuthService
import com.arcinteractive.spaces.data.mock.MockMembersRepository
import com.arcinteractive.spaces.data.model.canChangeRole
import com.arcinteractive.spaces.data.model.canRemove
import com.arcinteractive.spaces.data.model.hasCapability
import com.arcinteractive.spaces.data.model.roleFor
import com.arcinteractive.spaces.data.model.Space
import com.arcinteractive.spaces.data.model.SpaceInvite
import com.arcinteractive.spaces.data.model.SpaceMember
import com.arcinteractive.spaces.data.model.SpaceMemberRole
import com.arcinteractive.spaces.data.model.SpacePermission
import com.arcinteractive.spaces.data.spaces.SpaceService
import com.google.firebase.firestore.ListenerRegistration
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class MembersUiState(
    val members: List<SpaceMember>,
    val currentUserId: String? = null,
    val selectedMember: SpaceMember? = null,
    val isLoading: Boolean = false,
    val isCreatingInvite: Boolean = false,
    val pendingInvite: SpaceInvite? = null,
    val lastErrorMessage: String? = null
)

class MembersViewModel(
    space: Space,
    private val spaceService: SpaceService = SpaceService(),
    private val authService: AuthService = AuthService()
) : ViewModel() {
    val space: Space = space
    private var listener: ListenerRegistration? = null

    private val _uiState = MutableStateFlow(
        MembersUiState(members = MockMembersRepository.membersFor(space))
    )
    val uiState: StateFlow<MembersUiState> = _uiState.asStateFlow()

    fun loadMembersIfNeeded(context: Context) {
        if (listener != null) return
        _uiState.update {
            it.copy(
                isLoading = true,
                currentUserId = spaceService.currentUserId(context)
            )
        }
        listener = spaceService.listenToMembers(
            context = context,
            space = space,
            listenerKey = "members.${space.id}"
        ) { result ->
            result.onSuccess { members ->
                _uiState.update {
                    it.copy(
                        members = members,
                        currentUserId = spaceService.currentUserId(context),
                        isLoading = false
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        currentUserId = spaceService.currentUserId(context),
                        isLoading = false,
                        lastErrorMessage = error.localizedMessage ?: "Unable to load members."
                    )
                }
            }
        }
    }

    fun openMember(member: SpaceMember) {
        _uiState.update { it.copy(selectedMember = member) }
    }

    fun closeMember() {
        _uiState.update { it.copy(selectedMember = null) }
    }

    fun createInvite(context: Context) {
        openInviteEditor(context)
    }

    fun openInviteEditor(context: Context) {
        if (_uiState.value.isLoading || _uiState.value.isCreatingInvite) return

        _uiState.update { it.copy(isLoading = true, lastErrorMessage = null) }
        viewModelScope.launch {
            runCatching {
                spaceService.fetchLatestInvite(context, space)
            }.onSuccess { invite ->
                if (invite != null) {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            pendingInvite = invite
                        )
                    }
                } else {
                    _uiState.update { it.copy(isLoading = false) }
                    createFreshInvite(context)
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        lastErrorMessage = error.localizedMessage ?: "Unable to load the current invite link."
                    )
                }
            }
        }
    }

    private fun createFreshInvite(context: Context) {
        if (_uiState.value.isCreatingInvite) return
        _uiState.update { it.copy(isCreatingInvite = true, lastErrorMessage = null) }
        viewModelScope.launch {
            runCatching {
                spaceService.createInvite(context, space)
            }.onSuccess { invite ->
                _uiState.update {
                    it.copy(
                        isCreatingInvite = false,
                        pendingInvite = invite
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(
                        isCreatingInvite = false,
                        lastErrorMessage = error.localizedMessage ?: "Unable to create an invite."
                    )
                }
            }
        }
    }

    fun dismissInvite() {
        _uiState.update { it.copy(pendingInvite = null) }
    }

    fun updateSelectedMemberRole(context: Context, role: SpaceMemberRole) {
        val selectedMember = _uiState.value.selectedMember ?: return

        viewModelScope.launch {
            runCatching {
                spaceService.updateMemberRole(context, space, selectedMember.id, role)
            }.onSuccess {
                val updatedMember = selectedMember.copy(role = role)
                _uiState.update { state ->
                    state.copy(
                        members = state.members.map { member ->
                            if (member.id == updatedMember.id) updatedMember else member
                        },
                        selectedMember = updatedMember
                    )
                }
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to update this member.")
                }
            }
        }
    }

    fun removeSelectedMember(context: Context, onRemoved: () -> Unit) {
        val selectedMember = _uiState.value.selectedMember ?: return

        viewModelScope.launch {
            runCatching {
                spaceService.removeMember(context, space, selectedMember.id)
            }.onSuccess {
                _uiState.update { state ->
                    state.copy(
                        members = state.members.filterNot { it.id == selectedMember.id },
                        selectedMember = null
                    )
                }
                onRemoved()
            }.onFailure { error ->
                _uiState.update {
                    it.copy(lastErrorMessage = error.localizedMessage ?: "Unable to remove this member.")
                }
            }
        }
    }

    fun canInviteMembers(context: Context): Boolean {
        return space.hasCapability(_uiState.value.currentUserId, _uiState.value.members, SpacePermission.InviteMembers)
    }

    fun canManageRoles(context: Context): Boolean {
        return _uiState.value.members.any { canManageRoles(context, it) }
    }

    fun canManageMembers(context: Context): Boolean {
        return _uiState.value.members.any { canRemoveMember(context, it) }
    }

    fun canManageRoles(context: Context, member: SpaceMember): Boolean {
        return availableRoles(context, member).isNotEmpty()
    }

    fun canRemoveMember(context: Context, member: SpaceMember): Boolean {
        val currentRole = currentUserRole(context) ?: return false
        return currentRole.canRemove(member.role, member.id == space.ownerId)
    }

    fun availableRoles(context: Context, member: SpaceMember): List<SpaceMemberRole> {
        val currentRole = currentUserRole(context) ?: return emptyList()
        return SpaceMemberRole.entries.filter {
            currentRole.canChangeRole(member.role, it, member.id == space.ownerId)
        }
    }

    fun clearLastErrorMessage() {
        _uiState.update { it.copy(lastErrorMessage = null) }
    }

    override fun onCleared() {
        listener?.remove()
        listener = null
        super.onCleared()
    }

    private fun currentUserRole(context: Context): SpaceMemberRole? {
        return space.roleFor(_uiState.value.currentUserId, _uiState.value.members)
    }
}

File: iOS/Spaces V2/ViewModels/MembersViewModel.swift
import Combine
import FirebaseFirestore
import Foundation

@MainActor
final class MembersViewModel: ObservableObject {
    @Published private(set) var members: [SpaceMember]
    @Published private(set) var isLoading = false
    @Published private(set) var isCreatingInvite = false
    @Published private(set) var pendingInvite: SpaceInvite?
    @Published var errorMessage: String?

    let space: Space

    private let spaceService: SpaceService
    private let currentUserID: String?
    private var listener: ListenerRegistration?

    init(space: Space) {
        self.space = space
        self.members = MockData.spaceMembers(for: space)
        self.spaceService = SpaceService()
        self.currentUserID = AuthService().currentSession()?.uid
    }

    init(space: Space, members: [SpaceMember]? = nil, spaceService: SpaceService) {
        self.space = space
        self.members = members ?? MockData.spaceMembers(for: space)
        self.spaceService = spaceService
        self.currentUserID = AuthService().currentSession()?.uid
    }

    deinit {
        listener?.remove()
    }

    var canInviteMembers: Bool {
        space.hasCapability(.inviteMembers, for: currentUserID, members: members)
    }

    var canManageRoles: Bool {
        members.contains(where: { canManageRoles(for: $0) })
    }

    var canManageMembers: Bool {
        members.contains(where: { canRemove($0) })
    }

    func canManageRoles(for member: SpaceMember) -> Bool {
        !availableRoles(for: member).isEmpty
    }

    func availableRoles(for member: SpaceMember) -> [SpaceMemberRole] {
        guard let currentRole = currentUserRole else { return [] }
        return SpaceMemberRole.allCases.filter {
            currentRole.canChangeRole(
                of: member.role,
                to: $0,
                isTargetOwner: member.id == space.ownerId
            )
        }
    }

    func canRemove(_ member: SpaceMember) -> Bool {
        guard let currentRole = currentUserRole else { return false }
        return currentRole.canRemove(
            targetRole: member.role,
            isTargetOwner: member.id == space.ownerId
        )
    }

    func startListeningIfNeeded() {
        guard listener == nil else { return }
        isLoading = true
        listener = spaceService.listenToMembers(for: space) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let members):
                self.members = members
                self.isLoading = false
            case .failure(let error):
                self.errorMessage = error.localizedDescription
                self.isLoading = false
            }
        }
    }

    func createInvite() async {
        guard pendingInvite == nil else { return }
        await openInviteEditor()
    }

    func openInviteEditor() async {
        guard !isLoading else { return }
        isLoading = true
        defer { isLoading = false }

        do {
            if let existingInvite = try await spaceService.fetchLatestInvite(for: space) {
                pendingInvite = existingInvite
            } else {
                await createFreshInvite()
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func createFreshInvite() async {
        guard !isCreatingInvite else { return }
        isCreatingInvite = true
        defer { isCreatingInvite = false }

        do {
            pendingInvite = try await spaceService.createInvite(for: space)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func dismissInvite() {
        pendingInvite = nil
    }

    func clearError() {
        errorMessage = nil
    }

    func updateRole(for memberID: String, to role: SpaceMemberRole) async {
        do {
            try await spaceService.updateMemberRole(in: space, memberID: memberID, role: role)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func removeMember(withID memberID: String) async -> Bool {
        do {
            try await spaceService.removeMember(from: space, memberID: memberID)
            return true
        } catch {
            errorMessage = error.localizedDescription
            return false
        }
    }

    private var currentUserRole: SpaceMemberRole? {
        space.role(for: currentUserID, members: members)
    }
}
